File: src/lib/geo.js
// Worldwide place lookup. Nominatim geocodes anywhere on earth for free and
// returns names in the caller's language, so there is no place table to keep.
// ponytail: no API key, no proxy. Nominatim asks for <=1 req/sec, which the
// debounced search box already respects.

const NOMINATIM = 'https://nominatim.openstreetmap.org/search'
// The main Overpass instance returns 504 often enough to matter (it did while
// this was being written), so try the public mirrors before giving up.
const OVERPASS_MIRRORS = [
  'https://overpass-api.de/api/interpreter',
  'https://overpass.kumi.systems/api/interpreter',
  'https://overpass.osm.ch/api/interpreter'
]

// Street names do not change, so a place looked up once never needs Overpass
// again. This is what keeps a bad upstream from being visible at all.
const STREET_CACHE_PREFIX = 'roost-streets:'
const STREET_CACHE_TTL = 30 * 24 * 60 * 60 * 1000

function readStreetCache(id) {
  try {
    const raw = localStorage.getItem(STREET_CACHE_PREFIX + id)
    if (!raw) return null
    const { at, streets } = JSON.parse(raw)
    return Date.now() - at > STREET_CACHE_TTL ? null : streets
  } catch {
    return null
  }
}

function writeStreetCache(id, streets) {
  try {
    localStorage.setItem(STREET_CACHE_PREFIX + id, JSON.stringify({ at: Date.now(), streets }))
  } catch {
    // A full or blocked localStorage only costs us the cache, not the lookup.
  }
}

export const defaultPlace = {
  id: 'vancouver-bc',
  name: 'Vancouver',
  label: 'Vancouver, British Columbia, Canada',
  countryCode: 'CA',
  lat: 49.2827,
  lng: -123.1207,
  bbox: [49.198, -123.225, 49.317, -123.023]
}

function toPlace(r) {
  const [s, n, w, e] = r.boundingbox.map(Number)
  return {
    id: `${r.osm_type}-${r.osm_id}`,
    name: r.name || r.display_name.split(',')[0],
    label: r.display_name,
    countryCode: (r.address?.country_code || '').toUpperCase(),
    lat: Number(r.lat),
    lng: Number(r.lon),
    // Nominatim gives [south, north, west, east]; everything else here wants
    // [south, west, north, east].
    bbox: [s, w, n, e]
  }
}

export async function searchPlaces(query, language = 'en', signal) {
  if (!query.trim()) return []
  const url = `${NOMINATIM}?${new URLSearchParams({
    q: query,
    format: 'jsonv2',
    addressdetails: '1',
    limit: '8',
    'accept-language': language,
    featureType: 'settlement'
  })}`
  const res = await fetch(url, { signal, headers: { Accept: 'application/json' } })
  if (!res.ok) throw new Error(`Place search failed: ${res.status}`)
  const rows = await res.json()
  return rows.filter(r => r.boundingbox).map(toPlace)
}

// Real named streets inside the place, in the local script. Returns
// { streets, unavailable }: an empty list with unavailable=true means every
// mirror failed, which the UI says out loud instead of quietly inventing
// addresses at the city centroid.
export async function streetsNear(place, signal) {
  const cached = readStreetCache(place.id)
  if (cached) return { streets: cached, unavailable: false }

  const [s, w, n, e] = clampBox(place.bbox)
  const query = `[out:json][timeout:20];way["highway"~"^(residential|living_street|tertiary|secondary|unclassified)$"]["name"](${s},${w},${n},${e});out tags center 120;`

  for (const endpoint of OVERPASS_MIRRORS) {
    try {
      const res = await fetch(endpoint, { method: 'POST', body: query, signal })
      if (!res.ok) continue
      const data = await res.json()
      const seen = new Set()
      const streets = data.elements
        .filter(el => el.center && el.tags?.name && !seen.has(el.tags.name) && seen.add(el.tags.name))
        .map(el => ({ name: el.tags.name, lat: el.center.lat, lng: el.center.lon }))
      if (!streets.length) continue
      writeStreetCache(place.id, streets)
      return { streets, unavailable: false }
    } catch (err) {
      // An aborted request is the caller changing its mind, not an outage.
      if (err.name === 'AbortError') return { streets: [], unavailable: false }
    }
  }

  return { streets: [], unavailable: true }
}

// A country-sized bbox would ask Overpass for the whole nation. Cap the box at
// roughly a metro area around the centre.
function clampBox([s, w, n, e], maxDeg = 0.25) {
  const cy = (s + n) / 2
  const cx = (w + e) / 2
  return [
    Math.max(s, cy - maxDeg), Math.max(w, cx - maxDeg),
    Math.min(n, cy + maxDeg), Math.min(e, cx + maxDeg)
  ]
}

File: kmp/shared/src/commonMain/kotlin/com/nulljosh/roost/Geo.kt
package com.nulljosh.roost

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.parameter
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

data class Place(
    val id: String,
    val name: String,
    val label: String,
    val countryCode: String,
    val lat: Double,
    val lng: Double,
)

val DEFAULT_PLACE = Place(
    id = "vancouver-bc", name = "Vancouver", label = "Vancouver, British Columbia, Canada",
    countryCode = "CA", lat = 49.2827, lng = -123.1207,
)

@Serializable
private data class NominatimAddress(
    @SerialName("country_code") val countryCode: String? = null,
    val city: String? = null,
    val town: String? = null,
    val village: String? = null,
)

@Serializable
private data class NominatimResult(
    @SerialName("place_id") val placeId: Long,
    @SerialName("display_name") val displayName: String,
    val lat: String,
    val lon: String,
    val address: NominatimAddress? = null,
    val boundingbox: List<String>? = null,
)

// Ported from src/lib/geo.js's searchPlaces only. streetsNear (Overpass real
// street anchors) is not ported -- generateListings already has a fallback
// path for that case, so this just always takes it. See roadmap.md.
class GeoClient {
    private val http = HttpClient {
        install(ContentNegotiation) { json(Json { ignoreUnknownKeys = true }) }
    }

    suspend fun searchPlaces(query: String, language: String = "en"): List<Place> {
        if (query.isBlank()) return emptyList()
        val results: List<NominatimResult> = http.get("https://nominatim.openstreetmap.org/search") {
            parameter("q", query)
            parameter("format", "jsonv2")
            parameter("addressdetails", "1")
            parameter("limit", "8")
            parameter("accept-language", language)
            parameter("featureType", "settlement")
        }.body()
        return results.filter { it.boundingbox != null }.map { r ->
            val name = r.address?.city ?: r.address?.town ?: r.address?.village ?: r.displayName.substringBefore(",")
            Place(
                id = r.placeId.toString(),
                name = name,
                label = r.displayName,
                countryCode = (r.address?.countryCode ?: "").uppercase(),
                lat = r.lat.toDoubleOrNull() ?: 0.0,
                lng = r.lon.toDoubleOrNull() ?: 0.0,
            )
        }
    }
}
